File: frontend/src/services/messageService.js
import axiosInstance from '../utils/axios';

const DEFAULT_TIMEOUT = 15000;
const LONG_TIMEOUT = 30000;
const MAX_RETRIES = 2;

const validateUserId = (userId) => {
  if (!userId || typeof userId !== 'string' || userId.trim().length === 0) {
    throw new Error('Invalid user ID provided');
  }
  return userId.trim();
};

export const sendMessage = (sellerId, buyerId, productId, content) => {
  return axiosInstance.post('/api/messages', {
    sellerId,
    buyerId,
    productId,
    content
  }, { timeout: DEFAULT_TIMEOUT });
};

export const getSellerMessages = (sellerId) => {
  return axiosInstance.get(`/api/messages/seller/${sellerId}`, { timeout: DEFAULT_TIMEOUT });
};

export const getUnreadMessages = (sellerId) => {
  return axiosInstance.get(`/api/messages/unread/${sellerId}`);
};

export const markMessageAsRead = (messageId) => {
  return axiosInstance.put(`/api/messages/${messageId}/read`);
};

export const replyToMessage = (messageId, replyContent) => {
  return axiosInstance.post(`/api/messages/${messageId}/reply`, { replyContent });
};

export const getConversation = async (userId1, userId2, retryCount = 0) => {
  try {
    const response = await axiosInstance.get(
      `/api/messages/conversation/${userId1}/${userId2}`,
      { timeout: DEFAULT_TIMEOUT }
    );
    return response;
  } catch (error) {
    if (retryCount < MAX_RETRIES && error.code === 'ECONNABORTED') {
      await new Promise(resolve => setTimeout(resolve, 1000));
      return getConversation(userId1, userId2, retryCount + 1);
    }
    throw error;
  }
};

export const getUserMessages = async (userId, retryCount = 0) => {
  try {
    const validUserId = validateUserId(userId);
    const response = await axiosInstance.get(`/api/messages/user/${validUserId}`, {
      timeout: DEFAULT_TIMEOUT
    });
    return response;
  } catch (error) {
    if (retryCount < MAX_RETRIES && error.code === 'ECONNABORTED') {
      await new Promise(resolve => setTimeout(resolve, 1000));
      return getUserMessages(userId, retryCount + 1);
    }
    throw error;
  }
};

export const getGroupedConversations = async (userId, searchQuery = '') => {
  try {
    const validUserId = validateUserId(userId);
    const response = await axiosInstance.get(`/api/messages/conversations/${validUserId}`, {
      params: { search: searchQuery },
      timeout: DEFAULT_TIMEOUT
    });

    return { 
      data: response.data || [],
      status: response.status 
    };
  } catch (error) {
    if (error.response?.status === 404) {
      return { data: [], status: 404 };
    }
    console.error('Error fetching conversations:', error);
    return { data: [], status: error.response?.status || 500 };
  }
};

export const getBuyerMessages = (buyerId) => {
  return axiosInstance.get(`/api/messages/buyer/${buyerId}`);
};

File: backend/src/main/java/com/example/backend/service/MessageService.java
package com.example.backend.service;

import com.example.backend.model.Message;
import com.example.backend.model.User;
import com.example.backend.repository.MessageRepository;
import com.example.backend.repository.UserRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.time.LocalDateTime;
import java.util.*;
import java.util.logging.Level;
import java.util.logging.Logger;
import java.util.stream.Collectors;
import java.util.stream.Stream;

@Service
public class MessageService {
    private static final Logger logger = Logger.getLogger(MessageService.class.getName());

    @Autowired
    private MessageRepository messageRepository;

    @Autowired
    private UserRepository userRepository;

    public Message createMessage(String sellerId, String buyerId, String productId, String content) {
        Message message = new Message();
        message.setSellerId(sellerId);
        message.setBuyerId(buyerId);
        message.setSenderId(buyerId); // Set buyer as sender for initial message
        message.setReceiverId(sellerId); // Set seller as receiver for initial message
        message.setProductId(productId);
        message.setContent(content);
        message.setCreatedAt(LocalDateTime.now());
        message.setRead(false);

        Message savedMessage = messageRepository.save(message);
        List<Message> enriched = enrichMessagesWithUserNames(Collections.singletonList(savedMessage));
        return enriched.get(0);
    }

    public List<Message> getSellerMessages(String sellerId) {
        return messageRepository.findBySellerId(sellerId);
    }

    public List<Message> getBuyerMessages(String buyerId) {
        try {
            return messageRepository.findByBuyerIdOrderByCreatedAtDesc(buyerId);
        } catch (Exception e) {
            throw new RuntimeException("Failed to fetch buyer messages: " + e.getMessage());
        }
    }

    public List<Message> getUnreadMessages(String sellerId) {
        return messageRepository.findBySellerIdAndIsReadFalse(sellerId);
    }

    public Message markAsRead(String messageId) {
        Message message = messageRepository.findById(messageId)
            .orElseThrow(() -> new RuntimeException("Message not found"));
        message.setRead(true);
        return messageRepository.save(message);
    }

    public List<Message> getConversation(String userId1, String userId2) {
        if (userId1 == null || userId2 == null) {
            logger.warning("Invalid user IDs provided for conversation");
            return new ArrayList<>();
        }
        try {
            List<Message> messages = messageRepository.findMessagesByUsers(userId1.trim(), userId2.trim());
            List<Message> enriched = enrichMessagesWithUserNames(messages);
            return enriched.stream()
                .sorted(Comparator.comparing(Message::getCreatedAt))
                .collect(Collectors.toList());
        } catch (Exception e) {
            logger.log(Level.SEVERE, "Error fetching conversation: " + e.getMessage(), e);
            return new ArrayList<>();
        }
    }

    public List<Message> getUserMessages(String userId) {
        if (userId == null || userId.trim().isEmpty()) {
            throw new IllegalArgumentException("User ID cannot be null or empty");
        }

        try {
            List<Message> messages = messageRepository.findByBuyerIdOrSellerIdOrderByCreatedAtDesc(
                userId.trim(), userId.trim());

            if (messages == null) {
                return new ArrayList<>();
            }

            return enrichMessagesWithUserNames(messages);
        } catch (Exception e) {
            throw new RuntimeException("Error fetching user messages: " + e.getMessage());
        }
    }

    public Message replyToMessage(String originalMessageId, String content) {
        Message originalMessage = messageRepository.findById(originalMessageId)
            .orElseThrow(() -> new RuntimeException("Original message not found"));

        Message reply = new Message();
        reply.setSenderId(originalMessage.getReceiverId());
        reply.setReceiverId(originalMessage.getSenderId());
        reply.setSellerId(originalMessage.getSellerId());
        reply.setBuyerId(originalMessage.getBuyerId());
        reply.setProductId(originalMessage.getProductId());
        reply.setContent(content);
        reply.setCreatedAt(LocalDateTime.now());
        reply.setRead(false);

        return messageRepository.save(reply);
    }

    public List<Message> getMessageHistory(String userId) {
        List<Message> messages = messageRepository.findMessageHistory(userId);
        return enrichMessagesWithUserNames(messages);
    }

    public List<Message> getGroupedConversations(String userId) {
        if (userId == null || userId.trim().isEmpty()) {
            throw new IllegalArgumentException("User ID cannot be null or empty");
        }

        try {
            String trimmedUserId = userId.trim();
            List<Message> allMessages = messageRepository.findByBuyerIdOrSellerIdOrderByCreatedAtDesc(
                trimmedUserId, trimmedUserId);

            if (allMessages == null || allMessages.isEmpty()) {
                return new ArrayList<>();
            }

            Map<String, Message> latestMessages = new LinkedHashMap<>();

            for (Message message : allMessages) {
                if (message == null || message.getSenderId() == null || message.getReceiverId() == null) {
                    continue;
                }

                String partnerId;
                if (trimmedUserId.equals(message.getSenderId())) {
                    partnerId = message.getReceiverId();
                } else if (trimmedUserId.equals(message.getReceiverId())) {
                    partnerId = message.getSenderId();
                } else {
                    continue;
                }

                if (!latestMessages.containsKey(partnerId) ||
                    (message.getCreatedAt() != null &&
                     (latestMessages.get(partnerId).getCreatedAt() == null ||
                      message.getCreatedAt().isAfter(latestMessages.get(partnerId).getCreatedAt())))) {
                    latestMessages.put(partnerId, message);
                }
            }

            List<Message> conversations = new ArrayList<>(latestMessages.values());
            return enrichMessagesWithUserNames(conversations);
        } catch (Exception e) {
            throw new RuntimeException("Error processing conversations: " + e.getMessage(), e);
        }
    }

    private List<Message> enrichMessagesWithUserNames(List<Message> messages) {
        if (messages == null || messages.isEmpty()) {
            return new ArrayList<>();
        }

        Set<String> userIds = messages.stream()
            .flatMap(m -> Stream.of(m.getSenderId(), m.getReceiverId()))
            .filter(Objects::nonNull)
            .collect(Collectors.toSet());

        Map<String, User> userMap = userRepository.findAllById(userIds)
            .stream()
            .collect(Collectors.toMap(
                User::getId,
                user -> user,
                (u1, u2) -> u1
            ));

        return messages.stream()
            .filter(Objects::nonNull)
            .peek(message -> {
                User sender = userMap.get(message.getSenderId());
                User receiver = userMap.get(message.getReceiverId());
                message.setSenderName(sender != null ? sender.getFirstName() + " " + sender.getLastName() : "Unknown User");
                message.setReceiverName(receiver != null ? receiver.getFirstName() + " " + receiver.getLastName() : "Unknown User");
            })
            .collect(Collectors.toList());
    }
}
